muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java

```diff
@@ -2097,55 +2098,66 @@ public Response deleteWorkspaceStudent(@PathParam("WORKSPACEENTITYID") Long work
 
   @GET
   @Path("/workspaces/{WORKSPACEID}/journal")
-  @RESTPermitUnimplemented
+  @RESTPermit (handling = Handling.INLINE, requireLoggedIn = true)
   public Response listJournalEntries(
       @PathParam("WORKSPACEID") Long workspaceEntityId,
+      @QueryParam("userEntityId") Long userEntityId,
       @QueryParam("workspaceStudentId") String workspaceStudentId
   ) {
+    
+    // Workspace
+    
     List<WorkspaceJournalEntry> entries = new ArrayList<>();
     List<WorkspaceJournalEntryRESTModel> result = new ArrayList<>();
-    if (!sessionController.isLoggedIn()) {
-      return Response.status(Status.UNAUTHORIZED).build();
-    }
-    
     WorkspaceEntity workspaceEntity = workspaceController.findWorkspaceEntityById(workspaceEntityId);
     if (workspaceEntity == null) {
       return Response.status(Status.NOT_FOUND).build();
     }
+    
     UserEntity userEntity = sessionController.getLoggedUserEntity();
-    if (!sessionController.hasWorkspacePermission(MuikkuPermissions.LIST_ALL_JOURNAL_ENTRIES, workspaceEntity)) {
-      if (workspaceStudentId == null) {
-        entries = workspaceJournalController.listEntries(workspaceController.findWorkspaceEntityById(workspaceEntityId));
-      } else {
-        return Response.status(Status.UNAUTHORIZED).entity("Only teachers may look at others' journal entries").build();
+    boolean canListAllEntries = sessionController.hasWorkspacePermission(MuikkuPermissions.LIST_ALL_JOURNAL_ENTRIES, workspaceEntity);
+    if (workspaceStudentId == null && userEntityId == null && canListAllEntries) {
+      entries = workspaceJournalController.listEntries(workspaceController.findWorkspaceEntityById(workspaceEntityId));
+    }
+    else {
+      WorkspaceUserEntity workspaceUserEntity = null;
```

Comment:
workspaceuserentity can be removed from here and declared locally later on lines 2131, 2146